libeos-updater-util/avahi-service-file.c

```diff
@@ -37,67 +39,355 @@ const gchar * const EOS_UPDATER_AVAHI_SERVICE_TYPE = "_eos_updater._tcp";
+                                      guint8        *out_size_level,
+                                      GError       **error)
+{
+  guint8 size_level = EOS_OSTREE_AVAHI_SIZE_LEVEL_FIT_SINGLE_DNS_MESSAGE;
+
+  g_variant_dict_lookup (options_dict, EOS_OSTREE_AVAHI_OPTION_TXT_RECORDS_SIZE_LEVEL_Y,
+                         "y", &size_level);
+  switch (size_level)
+    {
+    case EOS_OSTREE_AVAHI_SIZE_LEVEL_CUSTOM:
+      if (!g_variant_dict_lookup (options_dict,
+                                  EOS_OSTREE_AVAHI_OPTION_TXT_RECORDS_CUSTOM_SIZE_T,
+                                  "t",
+                                  NULL))
+        {
+          g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                               "custom size level set, but no custom size "
+                               "limit passed to the options or it is of wrong type");
+          return FALSE;
+        }
+      /* fall through */
+
+    case EOS_OSTREE_AVAHI_SIZE_LEVEL_SUPPORT_CRAPPY_HARDWARE:
+    case EOS_OSTREE_AVAHI_SIZE_LEVEL_FIT_SINGLE_DNS_MESSAGE:
+    case EOS_OSTREE_AVAHI_SIZE_LEVEL_FIT_SINGLE_ETHERNET_PACKET:
+    case EOS_OSTREE_AVAHI_SIZE_LEVEL_FIT_SINGLE_MULTICAST_DNS_PACKET:
+    case EOS_OSTREE_AVAHI_SIZE_LEVEL_FIT_16_BIT_LIMIT:
+    case EOS_OSTREE_AVAHI_SIZE_LEVEL_ABSOLUTELY_LAX:
+      if (out_size_level)
+        *out_size_level = size_level;
+      return TRUE;
+
+    default:
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                   "unknown value %" G_GUINT16_FORMAT " for the %s option",
+                   (guint16) size_level,
+                   EOS_OSTREE_AVAHI_OPTION_TXT_RECORDS_SIZE_LEVEL_Y);
+      return FALSE;
+    }
+}
+
+static gboolean
+validate_total_size (gsize          total_size,
+                     GVariantDict  *options_dict,
+                     GError       **error)
+{
+  guint8 size_level;
+
+  if (!get_and_check_txt_records_size_level (options_dict, &size_level, error))
+    return FALSE;
+
+  switch (size_level)
+    {
+    case EOS_OSTREE_AVAHI_SIZE_LEVEL_CUSTOM:
+      {
+        guint64 custom_limit;
+        gboolean result = g_variant_dict_lookup (options_dict,
+                                                 EOS_OSTREE_AVAHI_OPTION_TXT_RECORDS_CUSTOM_SIZE_T,
+                                                 "t",
+                                                 &custom_limit);
+        g_assert (result);
+        if (total_size > custom_limit)
+          {
+            g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                         "txt records of size %" G_GSIZE_FORMAT
+                         " break the custom limit of %" G_GUINT64_FORMAT " bytes",
+                         total_size, custom_limit);
+            return FALSE;
+          }
+
+        break;
+      }
+
+    case EOS_OSTREE_AVAHI_SIZE_LEVEL_SUPPORT_CRAPPY_HARDWARE:
+      if (total_size > 256)
+        {
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
```

Comment:
No point in having separate error strings for all of these cases. You could put the size limits in a `const gsize[]` array indexed by the enum value, then factor out the error handling using a single error string like “TXT records of size %blah break the limit of %blah bytes.”.